### ISSUE TITLE
Active learning example requires that the initial sample size and query size match exactly

### ISSUE THREAD

**Issue by chschroeder** [recognai/rubrix#1622]
TL;DR: Small problem which causes the first batch to not be displayed, but is likely easy to fix.

**Describe the bug**

In the [active learning with small-text example](https://github.com/recognai/rubrix/blob/b6239fd474694cf4fe0c5f91a45cfcdb8d9f9ec7/docs/tutorials/active_learning_with_small_text.ipynb) I used a larger size of initial samples compared to the original version. Nothing happens in this case, i.e. I don't see a first batch in the UI. 

*Details:*
In the notebook we use the same size (`NUM_SAMPLES`) for initialization...

>
![num initial samples](https://user-images.githubusercontent.com/6340318/179820817-19f8df54-f2cf-4dd1-8c01-b4a8ad4bdd4f.png)

...and for querying:
![num query sizes](https://user-images.githubusercontent.com/6340318/179820991-a989a2f8-ff57-4bd3-b5d4-fb51bbb612e5.png)

For this reason the listener does not fire, since the condition checks for exactly `NUM_SAMPLES`-many samples:
`condition=lambda search: search.total==NUM_SAMPLES`.

*Suggestion:*
From the active learning perspective, it should not be a requirement that the initial sample size and query size must match. Consequently, there are three cases:

1. initial sample size > query_size
2. initial sample size == query_size
3. initial sample size < query_size

The second case already works. By changing the condition to `>=` I solved my specific problem (first case) as well. Now we still have to handle the third case. Honestly, I don't know how to fix this elegantly. I could treat the first batch separately, but that would also bloat the code. *Is there an additional clause we could add to the condition?*

**To Reproduce**
Navigate to the [active learning with small-text example](https://github.com/recognai/rubrix/blob/b6239fd474694cf4fe0c5f91a45cfcdb8d9f9ec7/docs/tutorials/active_learning_with_small_text.ipynb) and increase the initial number of samples (first picture), e.g. to 100.

**Expected behavior**
The initial data should be recognized as complete and the first active learning batch (batch-0) should be created and loaded into rubrix.

**Screenshots**
none

**Environment (please complete the following information):**
 - Rubrix Version 0.17.0.dev0
 - ElasticSearch Version 7.10.2

**Additional context**
none

**Comment by dcfidalgo** [comment]
Good catch @chschroeder, thanks for reporting! Your `>=` solution is not optimal, though, since if you annotate more than `NUM_SAMPLES` but not all of your batch 0, it will still go ahead and initialize the active learner with the labels available. The rest of batch 0 will get lost in Limbo, I guess ...
The easiest solution i can think of is moving the condition inside the `active_learning_loop` function, something like:
```python
@listener(
    dataset=DATASET_NAME,
    query="status:Validated AND metadata.batch_id:{batch_id}",
    execution_interval_in_seconds=3,
    batch_id=0
)
def active_learning_loop(records, ctx):
    condition = (ctx.query_params["batch_id"] == 0 and len(records) == INIT_SAMPLES) or len(records) == NUM_SAMPLES
    if not condition:
        return
    ...
```
and define a separate `INIT_SAMPLES` variable before.

@frascuchon Should we expose more info for the `condition` callable?

**Comment by frascuchon** [comment]
So, if I understand well, we're talking about 2 conditions:

* One for the initial batch
* One for the rest of the batches

It's that right? @chschroeder @dcfidalgo 

In that case, we could define the condition as @dcfidalgo mentions if we make accessible the listener query_params context inside the condition context. Something like that:

```python
def al_loop_step_completed(search, ctx):
   if ctx.query_params["batch_id"] == 0:
      return search.total == INIT_SAMPLES
   return search.total == NUM_SAMPLES

@listener(
    dataset=DATASET_NAME,
    query="status:Validated AND metadata.batch_id:{batch_id}",
    condition=al_loop_step_completed,
    execution_interval_in_seconds=3,
    batch_id=0
)
...

```

If this solution matches, I can easily apply the changes to resolve the problem.

**Comment by chschroeder** [comment]
@dcfidalgo Thanks for the quick response! Oh, I see, then this was just a workaround in this case ;).

@frascuchon Yes, that's correct. The main problem here is not the batches, but the circumstance that active learning has a chicken-and-egg problem: You need a model to query samples to obtain a (new) model.

An initially labeled batch is one way to solve this (i.e. to provide this initial model). If this initial batch should programmatically be treated otherwise is another question, but this is the only reason we need to make this distinction. I would guess we need some kind of distinction exactly for the use case above, where I want to initialize a model with 100 samples and query only a few samples (<100) per round or vice versa.

Your proposed solution with a function as condition looks good to me.

**Comment by frascuchon** [comment]
Okay, I'll work on that as soon as possible. I'll ping you when you can do some testing. Thanks !!!
